src/cpp/dds/domain/DomainParticipantImpl.cpp

```diff
@@ -135,38 +135,58 @@ DomainParticipantImpl::~DomainParticipantImpl()
 }
 
 
-bool DomainParticipantImpl::delete_publisher(
+ReturnCode_t DomainParticipantImpl::delete_publisher(
         Publisher* pub)
 {
+    if(participant_ != pub->get_participant())
+    {
+        return ReturnCode_t::RETCODE_PRECONDITION_NOT_MET;
+    }
     std::lock_guard<std::mutex> lock(mtx_pubs_);
     auto pit = publishers_.find(pub);
 
     if (pit != publishers_.end() && pub->get_instance_handle() == pit->second->get_instance_handle())
     {
+        std::vector<DataWriter*> writers;
+        pub->get_datawriters(writers);
+        if(!writers.empty())
+        {
+            return ReturnCode_t::RETCODE_PRECONDITION_NOT_MET;
+        }
         publishers_by_handle_.erase(publishers_by_handle_.find(pit->second->get_instance_handle()));
         delete pit->second;
         publishers_.erase(pit);
-        return true;
+        return ReturnCode_t::RETCODE_OK;
     }
 
-    return false;
+    return ReturnCode_t::RETCODE_ERROR;
 }
 
-bool DomainParticipantImpl::delete_subscriber(
+ReturnCode_t DomainParticipantImpl::delete_subscriber(
         Subscriber* sub)
 {
+    if(participant_ != sub->get_participant())
+    {
+        return ReturnCode_t::RETCODE_PRECONDITION_NOT_MET;
+    }
     std::lock_guard<std::mutex> lock(mtx_subs_);
     auto sit = subscribers_.find(sub);
 
     if (sit != subscribers_.end() && sub->get_instance_handle() == sit->second->get_instance_handle())
     {
+        std::vector<DataReader*> readers;
+        sub->get_datareaders(readers);
+        if(!readers.empty())
```

Comment:
Similar situation with `delete_publisher`.

```suggestion
        if (!readers.empty())
```

Reply:
Solved in PR #604